### ISSUE TITLE
Panic on too-short slice decode

### ISSUE THREAD

**Issue by elliotmr** [mitchellh/mapstructure#67]
I don't know if this is "expected behavior" or not, but after #61 was merged, Decoding a slice of structs will panic if given a pointer to a non-nil slice which is not long enough to hold the input data.

```
func TestNestedStructInSlice(t *testing.T) {
	t.Parallel()

	input := []map[string]interface{}{
		map[string]interface{}{
			"vstring": "foo",
			"vint":    42,
			"vbool":   true,
		},
		map[string]interface{}{
			"vstring": "bar",
			"vint":    52,
			"vbool":   false,
		},
	}

	result := []Basic{}
	err := Decode(input, &result)
	if err != nil {
		t.Fatalf("got an err: %s", err.Error())
	}
}
```

The issue is with the loop in at mapstructure.go:609. 

```
for i := 0; i < dataVal.Len(); i++ {
	currentData := dataVal.Index(i).Interface()
	currentField := valSlice.Index(i)

	fieldName := fmt.Sprintf("%s[%d]", name, i)
	if err := d.decode(fieldName, currentData, currentField); err != nil {
		errors = appendErrors(errors, err)
	}
}
```

If valSlice is shorter than dataVal we get a panic.

The question is: Should this throw an error, or automatically lengthen the input slice? I would guess the later would be appreciated (and that was the behavior before the pull request).

**Comment by sitano** [comment]
I've fixed this with 
```
> currentData := dataVal.Index(i).Interface()
		for valSlice.Len() <= i {
			valSlice = reflect.Append(valSlice, reflect.Zero(valElemType))
		}
> currentField := valSlice.Index(i)
```

It seems like this ough to be a part of logic of partial parse/merge due to line 571:
```
if valSlice.IsNil() || d.config.ZeroFields {
```
if ZeroFields is enabled, then the target slice will be recreated according to the valid length, or it will try to emerge new data into an existing slice otherwise.

**Comment by mitchellh** [comment]
Fixed by #68!
